modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1131 @@
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+{
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  std::vector<Real> angle(n_nodes, 0); // this assumes that the cutter mesh has compressed node id
+  std::vector<dof_id_type> node_id(_cut_elem_nnode);
+
+  const auto elem_begin = _cut_mesh->elements_begin();
+  auto elem_end = _cut_mesh->elements_end();
+  auto elem_it = elem_begin;
+
+  std::vector<Point> vertices(_cut_elem_nnode);
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+
+    vertices.clear();
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      vertices[i] = this_point;
+      node_id[i] = this_node->id();
+    }
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+      mooseAssert(node_id[i] < n_nodes, "Node ID is out of range");
+
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2](0), &vertices[0](0), &vertices[1](0));
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0](0), &vertices[1](0), &vertices[2](0));
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1](0), &vertices[2](0), &vertices[0](0));
+  }
+
+  // In each element, angles at three vertices are calculated.  Angles associated with all nodes are
+  // evaluated.
+  // Interior nodes will have a total angle = 2*pi; otherwise, it is a boundary node
+  // This assumes the cutter surface is flat.
+  auto node_begin = _cut_mesh->nodes_begin();
+  auto node_end = _cut_mesh->nodes_end();
+  auto node_it = node_begin;
+
+  for (; node_it != node_end; ++node_it)
+  {
+    Node * this_node = *node_it;
+    dof_id_type id = this_node->id();
+    if (!MooseUtils::relativeFuzzyEqual(angle[id], libMesh::pi * 2))
+    {
+      std::vector<dof_id_type> neighbors;
+      _boundary_map[id] = neighbors;
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::findBoundaryEdges()
+{
+  _boundary_edges.clear();
+
+  std::vector<dof_id_type> corner_elem_id;
+  const auto elem_begin = _cut_mesh->elements_begin();
+  auto elem_end = _cut_mesh->elements_end();
+  auto elem_it = elem_begin;
+  unsigned int counter = 0;
+
+  std::vector<dof_id_type> node_id(_cut_elem_nnode);
+  std::vector<bool> is_node_on_boundary(_cut_elem_nnode);
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    node_id.clear();
+    is_node_on_boundary.clear();
```

Comment:
Remove `node_id.clear()` and `is_node_on_boundary.clear()`.  Same deal as the other vector you were clearing.